support alibi

shcho1118 opened the pull request:
Like `causal` I added `alibi` support.
See `tests/test_alibi.py` for examples of usage.
In terms of performance (total # of tokens=16K), there is almost no loss by adding `alibi`.
`alibi` implementation is based on [here](https://github.com/ofirpress/attention_with_linear_biases/blob/a35aaca144e0eb6b789dfcb46784c4b8e31b7983/fairseq/models/transformer.py#L742)
test environment: A100 80GB, pytorch 2.0.1, cuda 11.7
![dim=2048, causal=True, forward](https://github.com/Dao-AILab/flash-attention/assets/20601334/92e6ba53-f351-4702-be0d-1e3d0425e9e0)
![dim=2048, causal=True, backward](https://github.com/Dao-AILab/flash-attention/assets/20601334/c5cffafd-c569-4b94-a4fd-4a24ac2276bb)
### TODO
- [x] add alibi in splitkv forward kernel
- [x] test varlen functions
- [x] simplify alibi parameters interface (e.g. local attention)
- [ ] implement symmetric alibi for non-causal case https://github.com/OpenGPTX/Megatron-LM/commit/50f961521d6a533f8dd4a1bc3bf9dc1fb7995513

tridao commented:
Thanks for the hard work! Does this also double compilation time?

shcho1118 commented:
command: `time python setup.py install`

`main` branch
real    9m35.418s
user    125m19.826s
sys     3m42.285s

`feat/alibi` branch
real    14m52.505s
user    196m36.594s
sys     4m30.820s

In my environment, it doesn't double, but does increase the time a bit.

UCC-team commented:
> command: `time python setup.py install`
> 
> `main` branch real 9m35.418s user 125m19.826s sys 3m42.285s
> 
> `feat/alibi` branch real 14m52.505s user 196m36.594s sys 4m30.820s
> 
> In my environment, it doesn't double, but increase the time a bit.
![image](https://github.com/Dao-AILab/flash-attention/assets/35629974/70a66380-0f95-4cd2-a394-71b3460c4db0)

![image](https://github.com/Dao-AILab/flash-attention/assets/35629974/0252384b-343f-4497-ac72-ae1189b26bf6)

![image](https://github.com/Dao-AILab/flash-attention/assets/35629974/cac51bda-dc4e-4740-8a34-6499a553ec43)

```
TypeError: fwd(): incompatible function arguments. The following argument types are supported:
E           1. (arg0: at::Tensor, arg1: at::Tensor, arg2: at::Tensor, arg3: Optional[at::Tensor], arg4: float, arg5: float, arg6: bool, arg7: bool, arg8: Optional[at::Generator]) -> List[at::Tensor]
```

shcho1118 commented:
> > command: `time python setup.py install`
> > `main` branch real 9m35.418s user 125m19.826s sys 3m42.285s
> > `feat/alibi` branch real 14m52.505s user 196m36.594s sys 4m30.820s
> > In my environment, it doesn't double, but increase the time a bit.
> > ![image](https://user-images.githubusercontent.com/35629974/269927888-70a66380-0f95-4cd2-a394-71b3460c4db0.png)
> 
> ![image](https://user-images.githubusercontent.com/35629974/269927697-0252384b-343f-4497-ac72-ae1189b26bf6.png)
> 
> ![image](https://user-images.githubusercontent.com/35629974/269927620-cac51bda-dc4e-4740-8a34-6499a553ec43.png)
> 
> ```
> TypeError: fwd(): incompatible function arguments. The following argument types are supported:
> E           1. (arg0: at::Tensor, arg1: at::Tensor, arg2: at::Tensor, arg3: Optional[at::Tensor], arg4: float, arg5: float, arg6: bool, arg7: bool, arg8: Optional[at::Generator]) -> List[at::Tensor]
> ```

Based on # of arguments of the error log you provided, I think you should uninstall previous flash attention (`pip uninstall flash-attn`) and reinstall it (`python setup.py install`).
If you use `feat/alibi` then # of arguments should be 12 not 9 since I added `alibi, alibi_start, alibi_ratio`.

c0g commented:
@shcho1118 sorry to nag, wondering if you'd consider updating this PR? It would be useful for my work. I can help resolve the merge conflicts if you like.

shcho1118 commented:
@c0g
I had other things to do and didn't pay attention to that PR.
I will resolve merge conflict as soon as possible.

c0g commented:
thank you :-)

shcho1118 commented:
@c0g
When I removed the merge conflict and ran `test_alibi.py` it, there was no error, so I think it should be fine.
@tridao 
Do you have any plans to merge this PR, or is there something I need to fix to make it work?

shcho1118 commented:
@c0g @tridao 
I found something that needed a quick fix.
We need to fix splitkv to use alibi even when it is used heuristically.

b-albar commented:
@shcho1118 Thanks for your work ! I've generalized your code a bit to adapt it to work without causal masking (see [here](https://github.com/Dao-AILab/flash-attention/compare/main...catie-aq:flash-attention:alibi-bias)). It implements the symetric option given in [this issue](https://github.com/ofirpress/attention_with_linear_biases/issues/5) and it should recover the usual alibi in case of causal masking. Maybe it could be useful for people like myself that still train encoder based architectures :)

jameswu2014 commented:
@shcho1118 flash_attn_varlen_func for forward, and the backward k and v will have Nan, How can I debug it?

shcho1118 commented:
@jameswu2014 

> @c0g @tridao I found something that needed a quick fix. We need to fix splitkv to use alibi even when it is used heuristically.

Maybe splitkv forward kernel was used heuristically since I didn't add alibi feature in that kernel yet.
I'm working on this issue.

jameswu2014 commented:
> @jameswu2014
> 
> > @c0g @tridao I found something that needed a quick fix. We need to fix splitkv to use alibi even when it is used heuristically.
> 
> Maybe splitkv forward kernel was used heuristically since I didn't add alibi feature in that kernel yet. I'm working on this issue. When you used varlen funcs before the last commit, did you have any problems?

I didn't use varlen funcs before. I tried it yesterday, and I got nan in dk/dv.

RonanKMcGovern commented:
Hi folks, any updates on merging this PR? thanks for your work on it.

shcho1118 commented:
@jameswu2014 (If you're still dealing with the same issue...)
I had the same problem you mentioned, and after adding the test case and trying it out, the problem seems to be solved.
The reason was that I had to add additional boundary checks inside `apply_alibi()`.

UCC-team commented:
I am using in Megatron. The forward execution is normal, but the backward execution reports an error and the input parameters are incorrect. What could be the reason？
FlashAttnVarlenFuncBackward：
<img width="1312" alt="image" src="https://github.com/Dao-AILab/flash-attention/assets/35629974/8bdefe49-1a74-4478-9385-7a939d3a6ce1">

shcho1118 commented:
@UCC-team 
We fixed the issue in a recent commit [c588fe4](https://github.com/Dao-AILab/flash-attention/pull/540/commits/c588fe437a969efa1ada53a3bcf3858ac781eef1)
Can you try uninstalling and reinstalling?
<img width="467" alt="스크린샷 2023-10-31 오전 10 44 26" src="https://github.com/Dao-AILab/flash-attention/assets/20601334/f34cb2f0-7cd9-42b5-8fb8-20fe24820b9f">
<img width="848" alt="스크린샷 2023-10-31 오전 10 44 46" src="https://github.com/Dao-AILab/flash-attention/assets/20601334/62f9bef7-7e20-405c-90c9-f118572a4c57">

jameswu2014 commented:
> @jameswu2014 (If you're still dealing with the same issue...) I had the same problem you mentioned, and after adding the test case and trying it out, the problem seems to be solved. The reason was that I had to add additional boundary checks inside `apply_alibi()`.

There is another question: In my case, my model's head num is 40 which is not the power of 2.So I cannot use your PR. Do you have any plan to support this kind of case? If you do so, You can only pass a alibi_slopes tensor rather than 3 parameters(alibi/alibi_start/alibi_ratio) to support it.

shcho1118 commented:
> > @jameswu2014 (If you're still dealing with the same issue...) I had the same problem you mentioned, and after adding the test case and trying it out, the problem seems to be solved. The reason was that I had to add additional boundary checks inside `apply_alibi()`.
> 
> There is another question: In my case, my model's head num is 40 which is not the power of 2.So I cannot use your PR. Do you have any plan to support this kind of case? If you do so, You can only pass a alibi_slopes tensor rather than 3 parameters(alibi/alibi_start/alibi_ratio) to support it.

I'm currently working on exactly what you mentioned.
This issue is ongoing, along with work on simplifying the interface like local attention.
Thank you for your frequent and helpful comments!!!

jameswu2014 commented:
> > > @jameswu2014 (If you're still dealing with the same issue...) I had the same problem you mentioned, and after adding the test case and trying it out, the problem seems to be solved. The reason was that I had to add additional boundary checks inside `apply_alibi()`.
> > 
> > 
> > There is another question: In my case, my model's head num is 40 which is not the power of 2.So I cannot use your PR. Do you have any plan to support this kind of case? If you do so, You can only pass a alibi_slopes tensor rather than 3 parameters(alibi/alibi_start/alibi_ratio) to support it.
> 
> I'm currently working on exactly what you mentioned. This issue is ongoing, along with work on simplifying the interface like local attention. Thank you for your frequent and helpful comments!!!

I have modified a version according to your implemention,maybe you can refer to it:https://github.com/jameswu2014/flash-attention/tree/feat/alibi

StarrySeas1 commented:
你好，请问你是基于原始仓库的那个提交进行修改的，以及主要做的那些修改。
`Require: Matrices Q, K, V ∈ R𝑁 ×𝑑 in HBM, on-chip SRAM of size 𝑀.
1: Setblocksizes𝐵 =𝑀,𝐵 =min 𝑀,𝑑. 𝑐4𝑑𝑟 4𝑑
2: InitializeO=(0)𝑁×𝑑 ∈R𝑁×𝑑,l=(0)𝑁 ∈R𝑁,𝑚=(−∞)𝑁 ∈R𝑁 inHBM.
3: DivideQinto𝑇 =l𝑁mblocksQ,...,Q ofsize𝐵 ×𝑑each,anddivideK,Vinto𝑇 =l𝑁mblocks
𝑟𝐵𝑟 1𝑇𝑟 𝑟 𝑐𝐵𝑐 K1,...,K𝑇𝑐 and V1,...,V𝑇𝑐, of size 𝐵𝑐 ×𝑑 each.
4: DivideOinto𝑇𝑟 blocksO𝑖,...,O𝑇𝑟 ofsize𝐵𝑟 ×𝑑each,dividelinto𝑇𝑟 blocksl𝑖,...,l𝑇𝑟 ofsize𝐵𝑟 each, divide 𝑚 into 𝑇𝑟 blocks 𝑚1,...,𝑚𝑇𝑟 of size 𝐵𝑟 each.
5: for 1 ≤ 𝑗 ≤ 𝑇𝑐 do
6: Load K𝑗,V𝑗 from HBM to on-chip SRAM.
7: for1≤𝑖≤𝑇𝑟 do
8: Load Q𝑖,O𝑖,l𝑖,𝑚𝑖 from HBM to on-chip SRAM.
9: Onchip,computeS𝑖𝑗 =Q𝑖K𝑇𝑗 ∈R𝐵𝑟×𝐵𝑐.
10: On chip, compute 𝑚 ̃ = rowmax(S ) ∈ R𝐵𝑟 , P ̃ = exp(S − 𝑚 ̃ ) ∈ R𝐵𝑟 ×𝐵𝑐 (pointwise), l ̃ = 𝑖𝑗𝑖𝑗𝑖𝑗𝑖𝑗𝑖𝑗 𝑖𝑗
rowsum(P ̃𝑖𝑗) ∈ R𝐵𝑟 .
11: Onchip,compute𝑚𝑖 =max(𝑚𝑖,𝑚 ̃𝑖𝑗)∈R ,l𝑖 =𝑒 𝑖 l𝑖+𝑒 𝑖 l𝑖𝑗 ∈R .
new −1 𝑚𝑖−𝑚new
12: WriteO𝑖 ←diag(l𝑖 ) (diag(l𝑖)𝑒 𝑖  是在这一步进行的添加吗？可以写下计算过程吗？
13: Write l𝑖 ← lnew, 𝑚𝑖 ← 𝑚new to HBM. 𝑖𝑖
14: end for
15: end for
16: Return O.
`

shcho1118 commented:
> Hello, may I ask which modifications you made based on the original code(which commit id), and which major modifications you made? Below are the calculation steps in the original paper. Did you add them in step 12?
> 
> ```batchfile
> Require: Matrices Q, K, V ∈ R𝑁 ×𝑑 in HBM, on-chip SRAM of size 𝑀.
> 1: Setblocksizes𝐵 =�𝑀�,𝐵 =min��𝑀�,𝑑�. 𝑐4𝑑𝑟 4𝑑
> 2: InitializeO=(0)𝑁×𝑑 ∈R𝑁×𝑑,l=(0)𝑁 ∈R𝑁,𝑚=(−∞)𝑁 ∈R𝑁 inHBM.
> 3: DivideQinto𝑇 =l𝑁mblocksQ,...,Q ofsize𝐵 ×𝑑each,anddivideK,Vinto𝑇 =l𝑁mblocks
> 𝑟𝐵𝑟 1𝑇𝑟 𝑟 𝑐𝐵𝑐 K1,...,K𝑇𝑐 and V1,...,V𝑇𝑐, of size 𝐵𝑐 ×𝑑 each.
> 4: DivideOinto𝑇𝑟 blocksO𝑖,...,O𝑇𝑟 ofsize𝐵𝑟 ×𝑑each,dividelinto𝑇𝑟 blocksl𝑖,...,l𝑇𝑟 ofsize𝐵𝑟 each, divide 𝑚 into 𝑇𝑟 blocks 𝑚1,...,𝑚𝑇𝑟 of size 𝐵𝑟 each.
> 5: for 1 ≤ 𝑗 ≤ 𝑇𝑐 do
> 6: Load K𝑗,V𝑗 from HBM to on-chip SRAM.
> 7: for1≤𝑖≤𝑇𝑟 do
> 8: Load Q𝑖,O𝑖,l𝑖,𝑚𝑖 from HBM to on-chip SRAM.
> 9: Onchip,computeS𝑖𝑗 =Q𝑖K𝑇𝑗 ∈R𝐵𝑟×𝐵𝑐.
> 10: On chip, compute 𝑚 ̃ = rowmax(S ) ∈ R𝐵𝑟 , P ̃ = exp(S − 𝑚 ̃ ) ∈ R𝐵𝑟 ×𝐵𝑐 (pointwise), l ̃ = 𝑖𝑗𝑖𝑗𝑖𝑗𝑖𝑗𝑖𝑗 𝑖𝑗
> rowsum(P ̃𝑖𝑗) ∈ R𝐵𝑟 .
> 11: Onchip,compute𝑚𝑖 =max(𝑚𝑖,𝑚 ̃𝑖𝑗)∈R ,l𝑖 =𝑒 𝑖 l𝑖+𝑒 𝑖 l𝑖𝑗 ∈R .
> new −1 𝑚𝑖−𝑚new
> 12: WriteO𝑖 ←diag(l𝑖 ) (diag(l𝑖)𝑒 𝑖  ## Was it added in this step? Can you write down the calculation?
> 13: Write l𝑖 ← lnew, 𝑚𝑖 ← 𝑚new to HBM. 𝑖𝑖
> 14: end for
> 15: end for
> 16: Return O.
> ```

After step 9, I generate alibi (alibi shape == Sij shape) on-chip & Sij += alibi, so no alibi tensor load from HBM to on-chip SRAM.

yoavhacohen commented:
Is it possible to support for 2D inputs?

StarrySeas1 commented:
> > Hello, may I ask which modifications you made based on the original code(which commit id), and which major modifications you made? Below are the calculation steps in the original paper. Did you add them in step 12?
> > ```batchfile
> > Require: Matrices Q, K, V ∈ R𝑁 ×𝑑 in HBM, on-chip SRAM of size 𝑀.
> > 1: Setblocksizes𝐵 =�𝑀�,𝐵 =min��𝑀�,𝑑�. 𝑐4𝑑𝑟 4𝑑
> > 2: InitializeO=(0)𝑁×𝑑 ∈R𝑁×𝑑,l=(0)𝑁 ∈R𝑁,𝑚=(−∞)𝑁 ∈R𝑁 inHBM.
> > 3: DivideQinto𝑇 =l𝑁mblocksQ,...,Q ofsize𝐵 ×𝑑each,anddivideK,Vinto𝑇 =l𝑁mblocks
> > 𝑟𝐵𝑟 1𝑇𝑟 𝑟 𝑐𝐵𝑐 K1,...,K𝑇𝑐 and V1,...,V𝑇𝑐, of size 𝐵𝑐 ×𝑑 each.
> > 4: DivideOinto𝑇𝑟 blocksO𝑖,...,O𝑇𝑟 ofsize𝐵𝑟 ×𝑑each,dividelinto𝑇𝑟 blocksl𝑖,...,l𝑇𝑟 ofsize𝐵𝑟 each, divide 𝑚 into 𝑇𝑟 blocks 𝑚1,...,𝑚𝑇𝑟 of size 𝐵𝑟 each.
> > 5: for 1 ≤ 𝑗 ≤ 𝑇𝑐 do
> > 6: Load K𝑗,V𝑗 from HBM to on-chip SRAM.
> > 7: for1≤𝑖≤𝑇𝑟 do
> > 8: Load Q𝑖,O𝑖,l𝑖,𝑚𝑖 from HBM to on-chip SRAM.
> > 9: Onchip,computeS𝑖𝑗 =Q𝑖K𝑇𝑗 ∈R𝐵𝑟×𝐵𝑐.
> > 10: On chip, compute 𝑚 ̃ = rowmax(S ) ∈ R𝐵𝑟 , P ̃ = exp(S − 𝑚 ̃ ) ∈ R𝐵𝑟 ×𝐵𝑐 (pointwise), l ̃ = 𝑖𝑗𝑖𝑗𝑖𝑗𝑖𝑗𝑖𝑗 𝑖𝑗
> > rowsum(P ̃𝑖𝑗) ∈ R𝐵𝑟 .
> > 11: Onchip,compute𝑚𝑖 =max(𝑚𝑖,𝑚 ̃𝑖𝑗)∈R ,l𝑖 =𝑒 𝑖 l𝑖+𝑒 𝑖 l𝑖𝑗 ∈R .
> > new −1 𝑚𝑖−𝑚new
> > 12: WriteO𝑖 ←diag(l𝑖 ) (diag(l𝑖)𝑒 𝑖  ## Was it added in this step? Can you write down the calculation?
> > 13: Write l𝑖 ← lnew, 𝑚𝑖 ← 𝑚new to HBM. 𝑖𝑖
> > 14: end for
> > 15: end for
> > 16: Return O.
> > ```
> 
> After step 9, I generate alibi (alibi shape == Sij shape) on-chip & Sij += alibi, so no alibi tensor load from HBM to on-chip SRAM.

1. Can you tell me which file this implementation code is in? (I generate alibi (alibi shape == Sij shape) on-chip & Sij += alibi, so no alibi tensor load from HBM to on-chip SRAM.)
2. Most of the calculation logic in the source code is encapsulated in flash_attn_2_cuda, but this part of the code is not found in the flash attention code base. Do you know where this library is? Thanks.

StarrySeas1 commented:
> Is it possible to support for 2D inputs?

The three matrices Q, K and V are two-dimensional.

shcho1118 commented:
> > > Hello, may I ask which modifications you made based on the original code(which commit id), and which major modifications you made? Below are the calculation steps in the original paper. Did you add them in step 12?
> > > ```batchfile
> > > Require: Matrices Q, K, V ∈ R𝑁 ×𝑑 in HBM, on-chip SRAM of size 𝑀.
> > > 1: Setblocksizes𝐵 =�𝑀�,𝐵 =min��𝑀�,𝑑�. 𝑐4𝑑𝑟 4𝑑
> > > 2: InitializeO=(0)𝑁×𝑑 ∈R𝑁×𝑑,l=(0)𝑁 ∈R𝑁,𝑚=(−∞)𝑁 ∈R𝑁 inHBM.
> > > 3: DivideQinto𝑇 =l𝑁mblocksQ,...,Q ofsize𝐵 ×𝑑each,anddivideK,Vinto𝑇 =l𝑁mblocks
> > > 𝑟𝐵𝑟 1𝑇𝑟 𝑟 𝑐𝐵𝑐 K1,...,K𝑇𝑐 and V1,...,V𝑇𝑐, of size 𝐵𝑐 ×𝑑 each.
> > > 4: DivideOinto𝑇𝑟 blocksO𝑖,...,O𝑇𝑟 ofsize𝐵𝑟 ×𝑑each,dividelinto𝑇𝑟 blocksl𝑖,...,l𝑇𝑟 ofsize𝐵𝑟 each, divide 𝑚 into 𝑇𝑟 blocks 𝑚1,...,𝑚𝑇𝑟 of size 𝐵𝑟 each.
> > > 5: for 1 ≤ 𝑗 ≤ 𝑇𝑐 do
> > > 6: Load K𝑗,V𝑗 from HBM to on-chip SRAM.
> > > 7: for1≤𝑖≤𝑇𝑟 do
> > > 8: Load Q𝑖,O𝑖,l𝑖,𝑚𝑖 from HBM to on-chip SRAM.
> > > 9: Onchip,computeS𝑖𝑗 =Q𝑖K𝑇𝑗 ∈R𝐵𝑟×𝐵𝑐.
> > > 10: On chip, compute 𝑚 ̃ = rowmax(S ) ∈ R𝐵𝑟 , P ̃ = exp(S − 𝑚 ̃ ) ∈ R𝐵𝑟 ×𝐵𝑐 (pointwise), l ̃ = 𝑖𝑗𝑖𝑗𝑖𝑗𝑖𝑗𝑖𝑗 𝑖𝑗
> > > rowsum(P ̃𝑖𝑗) ∈ R𝐵𝑟 .
> > > 11: Onchip,compute𝑚𝑖 =max(𝑚𝑖,𝑚 ̃𝑖𝑗)∈R ,l𝑖 =𝑒 𝑖 l𝑖+𝑒 𝑖 l𝑖𝑗 ∈R .
> > > new −1 𝑚𝑖−𝑚new
> > > 12: WriteO𝑖 ←diag(l𝑖 ) (diag(l𝑖)𝑒 𝑖  ## Was it added in this step? Can you write down the calculation?
> > > 13: Write l𝑖 ← lnew, 𝑚𝑖 ← 𝑚new to HBM. 𝑖𝑖
> > > 14: end for
> > > 15: end for
> > > 16: Return O.
> > > ```
> > 
> > 
> > After step 9, I generate alibi (alibi shape == Sij shape) on-chip & Sij += alibi, so no alibi tensor load from HBM to on-chip SRAM.
> 
> 1. Can you tell me which file this implementation code is in? (I generate alibi (alibi shape == Sij shape) on-chip & Sij += alibi, so no alibi tensor load from HBM to on-chip SRAM.)
> 2. Most of the calculation logic in the source code is encapsulated in flash_attn_2_cuda, but this part of the code is not found in the flash attention code base. Do you know where this library is? Thanks.

Please check `apply_alibi()` inside `csrc/flash_attn/src/alibi.h` and how to use it. It is easy to understand since alibi logic itself is rather simple.

LyricZhao commented:
Hi, is there a timeline to merge this PR? Really need it, thanks for your hard work :)

LyricZhao commented:
> Hi, is there a timeline to merge this PR? Really need it, thanks for your hard work :)

Also, this PR seems overlapped with: https://github.com/Dao-AILab/flash-attention/pull/617. So I am wondering what is the plan/order for merging these 2 PRs to main.

vchiley commented:
> > Hi, is there a timeline to merge this PR? Really need it, thanks for your hard work :)
> 
> Also, this PR seems overlapped with: #617. So I am wondering what is the plan/order for merging these 2 PRs to main.

https://github.com/Dao-AILab/flash-attention/pull/617 adds a bias term to the kernel which can be used for alibi
This PR integrates the ALiBi positional embedding into the kernel itself; this should run faster than integrating a bias into the kernel.

Hi, is there a timeline to merge this PR?
^ this question still stands

tridao commented:
Thanks @shcho1118. Sorry this is taking me a while, I've been swamped lately.
This weekend or next week I might be able to merge.

shcho1118 commented:
@ofirpress asked for a performance comparison and wanted to share the results, which are shown below.
Rotary was measured by including `flash_rotary()` below in `benchmarks/benchmark_flash_attention.py`.
`flash_rotary()` uses flash attention triton kernel `apply_rotary_emb()` for rotary embedding so much faster than torch ver.
Still, there seems to be some performance difference when `head_dim` is large.
![a100 80g, cuda 11 7, pytorch 2 0 1, fwd](https://github.com/Dao-AILab/flash-attention/assets/20601334/81574dbf-21f2-4bae-8bed-1cee59a621f3)
![a100 80g, cuda 11 7, pytorch 2 0 1, bwd](https://github.com/Dao-AILab/flash-attention/assets/20601334/cc448219-34c6-4810-b6ee-683d85be50fd)
```
def flash_rotary(q, k, v, causal=False):

    def generate_cos_sin(seqlen, rotary_dim, device, dtype):
        assert rotary_dim % 2 == 0
        angle = torch.rand(seqlen * 2, rotary_dim // 2, device=device) * 2 * math.pi
        cos = torch.cos(angle).to(dtype=dtype)
        sin = torch.sin(angle).to(dtype=dtype)
        return cos, sin

    headdim_q = q.shape[-1]
    rotary_dim = int(1.0 * headdim_q)
    cos, sin = generate_cos_sin(q.shape[1], rotary_dim, "cuda", q.dtype)
    q = apply_rotary_emb(
        q, cos, sin, seqlen_offsets=0, interleaved=False, inplace=False
    )
    k = apply_rotary_emb(
        k, cos, sin, seqlen_offsets=0, interleaved=False, inplace=False
    )
    
    return flash_attn_func(q, k, v, causal=causal)
```

tridao commented:
I think the cos and sin generation should be cached and not counted as part of benchmark.

shcho1118 commented:
> I think the cos and sin generation should be cached and not counted as part of benchmark. Also for rotary I use 
`inplace=True`, idk if that makes any speed difference (it does save memory).

Thanks for the advice!
When I implemented the changes you mentioned (cos, sin caching, inplace=True), the result is as follows.
It doesn't seem to show much of a difference in trend from the previous results.
![a100 80g, cuda 11 7, pytorch 2 0 1, fwd, fixed](https://github.com/Dao-AILab/flash-attention/assets/20601334/243d9fd7-2aa1-4b65-a0af-d112a3835deb)
![a100 80g, cuda 11 7, pytorch 2 0 1, bwd, fixed](https://github.com/Dao-AILab/flash-attention/assets/20601334/39799ad6-e922-45f5-aec8-ad6dd6efef88)

vchiley commented:
Instead of passing in `alibi_start` and `alibi_ratio` to compute `alibi_slope`, can the ui be modified such that a user can pass in a tensor of `alibi_slopes` (1 per head). This will be useful for testing seq len extrapolation methods with ALiBi and is generally the ui in other attention implementations which support ALiBi (specifically inference libraries).

veya2ztn commented:
The default [alibi implement](https://github.com/ofirpress/attention_with_linear_biases/blob/a35aaca144e0eb6b789dfcb46784c4b8e31b7983/fairseq/models/transformer.py#L742) ignore the relative position between row and col, and only use the col index for alibi. The position index looks like (slen=10)
```
[[0 1 2 3 4 5 6 7 8 9]
 [0 1 2 3 4 5 6 7 8 9]
 [0 1 2 3 4 5 6 7 8 9]
 [0 1 2 3 4 5 6 7 8 9]
 [0 1 2 3 4 5 6 7 8 9]
 [0 1 2 3 4 5 6 7 8 9]
 [0 1 2 3 4 5 6 7 8 9]
 [0 1 2 3 4 5 6 7 8 9]
 [0 1 2 3 4 5 6 7 8 9]
 [0 1 2 3 4 5 6 7 8 9]]
```


 For those `non-cacual` case Alibi implement like it in bert ([Here ](https://huggingface.co/jinaai/jina-bert-implementation/blob/main/modeling_bert.py) is an example), the alibi bias added along the `relative offset =abs(row_idx-col_idx)`.   See 
```
context_position = torch.arange(size, device=device)[:, None]
memory_position = torch.arange(size, device=device)[None, :]
relative_position = torch.abs(memory_position - context_position) 
# [n_heads, max_token_length, max_token_length]
relative_position = relative_position.unsqueeze(0).expand(n_heads, -1, -1)
slopes = torch.Tensor(_get_alibi_head_slopes(n_heads)).to(device) * -1
alibi = slopes.unsqueeze(1).unsqueeze(1) * relative_position
```
The `relative_position` is (for slen=10) 
```
[[0 1 2 3 4 5 6 7 8 9]
 [1 0 1 2 3 4 5 6 7 8]
 [2 1 0 1 2 3 4 5 6 7]
 [3 2 1 0 1 2 3 4 5 6]
 [4 3 2 1 0 1 2 3 4 5]
 [5 4 3 2 1 0 1 2 3 4]
 [6 5 4 3 2 1 0 1 2 3]
 [7 6 5 4 3 2 1 0 1 2]
 [8 7 6 5 4 3 2 1 0 1]
 [9 8 7 6 5 4 3 2 1 0]]
```
Maybe it can be considered as an extra option for the build-in feature of `flash-attention-alibi` function.

I simple modify `alibi.h` as follow
```
#include <cmath>

#include <cute/tensor.hpp>

#include <cutlass/cutlass.h>
#include <cutlass/array.h>

#include "utils.h"

namespace flash {

using namespace cute;

////////////////////////////////////////////////////////////////////////////////////////////////////

template <typename Engine, typename Layout>
inline __device__ void apply_alibi(Tensor<Engine, Layout> &tensor, 
                                   const int col_idx_offset_,
                                   const int max_seqlen_k, 
                                   const int row_idx_offset_,
                                   const int max_seqlen_q, 
                                   const int warp_row_stride,
                                   const int head_idx,
                                   const float softmax_scale,
                                   const float alibi_start,
                                   const float alibi_ratio) {
    // tensor has shape (ncol=(2, MMA_M), nrow=(2, MMA_N))
    static_assert(Layout::rank == 2, "Only support 2D Tensor");
    const int lane_id = threadIdx.x % 32;
    const int row_idx_offset = row_idx_offset_;
    const int col_idx_offset = col_idx_offset_ + (lane_id % 4) * 2;
    const float alibi_slope = alibi_start * powf(alibi_ratio, (float)head_idx);
    int offset;
    float alibi;
    #pragma unroll
    for (int mi = 0; mi < size<0, 1>(tensor); ++mi) {
        const int row_idx_base = row_idx_offset + mi * warp_row_stride;
        #pragma unroll
        for (int i = 0; i < size<0, 0>(tensor); ++i) {
            const int row_idx = row_idx_base + i * 8;
            #pragma unroll
            for (int nj = 0; nj < size<1, 1>(tensor); ++nj) {
                const int col_idx_base = col_idx_offset + nj * 8;
                #pragma unroll
                for (int j = 0; j < size<1, 0>(tensor); ++j) {
                    const int col_idx = col_idx_base + j;
                    if (col_idx < row_idx)
                    {
                        offset =  (row_idx - col_idx) ; // <--- just add relative offset
                    }
                    else{
                        offset =  (col_idx - row_idx) ;
                    }
                    alibi = (alibi_slope * offset) / softmax_scale;
                    // const float alibi = (alibi_slope * col_idx) / softmax_scale;
                    if (col_idx < max_seqlen_k && row_idx < max_seqlen_q) {
                        tensor(make_coord(i, mi), make_coord(j, nj)) += alibi;
                    }
                }
            }
        }
    }
}

}  // namespace flash
```

It works. I suggest to add an arguement such as `position_mode` to switch between those two modes.

shcho1118 commented:
@veya2ztn I've implemented what you mentioned before, but had to undo commit due to performance issues.
The performance issue was for a different reason, and I will implement loading alibi_slope and then push it back.
Also, it is on TODO list :)

shcho1118 commented:
> Instead of passing in `alibi_start` and `alibi_ratio` to compute `alibi_slope`, can the ui be modified such that a user can pass in a tensor of `alibi_slopes` (1 per head). This will be useful for testing seq len extrapolation methods with ALiBi and is generally the ui in other attention implementations which support ALiBi (specifically inference libraries).

@vchiley 
I took your advice and implemented `alibi_slopes` loading.
However, I changed the shape to `(batch_size, num_heads)` instead of `(num_heads)`.
This is because when applying techniques like dynamic positional interpolation in inference (I don't know if it's widely used, but I've seen cases), slopes can vary from seqeunce to seqeunce.
When using the normal alibi, you can just use `(num_heads)` -> `(batch_size, num_heads)` repeat.
Only the last `stride` needs to be 1, so you don't even need to call `.contiguous()`.

tridao commented on csrc/flash_attn/src/flash_fwd_launch_template.h:
```diff
@@ -142,7 +144,9 @@ void run_mha_fwd_hdim32(Flash_fwd_params &params, cudaStream_t stream) {
     BOOL_SWITCH(params.p_dropout < 1.f, Is_dropout, [&] {
         BOOL_SWITCH(params.is_causal, Is_causal, [&] {
-            run_flash_fwd<Flash_fwd_kernel_traits<Headdim, 128, 128, 4, false, false, T>, Is_dropout, Is_causal>(params, stream);
+            BOOL_SWITCH(params.is_alibi, Is_alibi, [&] {
```
If you're already templating in "run_flash_fwd", you don't need to template here (and the rest of the file).

tridao commented on csrc/flash_attn/src/flash_bwd_launch_template.h:
```diff
@@ -18,20 +18,20 @@ __global__ void flash_bwd_clear_dkvaccum_kernel(Flash_bwd_params params) {
 }
 
-template<typename Kernel_traits, bool Is_dropout, bool Is_causal, bool Is_even_M, bool Is_even_K>
+template<typename Kernel_traits, bool Is_dropout, bool Is_causal, bool Is_alibi, bool Is_even_M, bool Is_even_K>
```
Let's call it "Has_alibi" instead of "Is_alibi"? Since this stands for "has attention linear bias".

tridao commented:
Thanks again @shcho1118! I just have a few stylistic comments. If you can rebase on main then I can merge.

shcho1118 commented:
> Thanks again @shcho1118! I just have a few stylistic comments. If you can rebase on main then I can merge.

Thanks @tridao.
I'll try to fix it as soon as I can.
Also, I'd like you to mention @kakaobrain when you merge because the work I did on this PR was driven by internal demand.

shcho1118 commented:
@tridao 
I've rebased on main branch and made changes based on the comments.
I think it's safe to merge now.

beginlner commented:
Hi, @shcho1118 , thanks for your contribution. I noticed a minor issue when utilizing GQA in the decoding process.
https://github.com/Dao-AILab/flash-attention/blob/e3a6941f051b4369a2862b555691deab2afedf86/csrc/flash_attn/flash_api.cpp#L1169

shcho1118 commented:
> Hi, @shcho1118 , thanks for your contribution. I noticed a minor issue when using GQA during decoding. Reshaped `q` leads to incorrect results.
> 
> https://github.com/Dao-AILab/flash-attention/blob/e3a6941f051b4369a2862b555691deab2afedf86/csrc/flash_attn/flash_api.cpp#L1169

Thank you for reporting this issue.
I think it's probably fixed by just adding
`seqlenq_ngroups_swapped = seqlenq_ngroups_swapped && !alibi_slopes_.has_value()`
I can't test it right now, so I'll work on it as soon as I can.

beginlner commented:
> > Hi, @shcho1118 , thanks for your contribution. I noticed a minor issue when using GQA during decoding. Reshaped `q` leads to incorrect results.
> > https://github.com/Dao-AILab/flash-attention/blob/e3a6941f051b4369a2862b555691deab2afedf86/csrc/flash_attn/flash_api.cpp#L1169
> 
> Thank you for reporting this issue. I think it's probably fixed by just adding `seqlenq_ngroups_swapped = seqlenq_ngroups_swapped && !alibi_slopes_.has_value()` I can't test it right now, so I'll work on it as soon as I can.

Maybe we can support both `seqlenq_ngroups_swapped` and `alibi_slopes` by check `causal` in the kernel like this:
https://github.com/Dao-AILab/flash-attention/blob/main/csrc/flash_attn/src/flash_fwd_kernel.h#L862

shcho1118 commented:
> > > Hi, @shcho1118 , thanks for your contribution. I noticed a minor issue when using GQA during decoding. Reshaped `q` leads to incorrect results.
> > > https://github.com/Dao-AILab/flash-attention/blob/e3a6941f051b4369a2862b555691deab2afedf86/csrc/flash_attn/flash_api.cpp#L1169
> > 
> > 
> > Thank you for reporting this issue. I think it's probably fixed by just adding `seqlenq_ngroups_swapped = seqlenq_ngroups_swapped && !alibi_slopes_.has_value()` I can't test it right now, so I'll work on it as soon as I can.
> 
> Maybe we can support both `seqlenq_ngroups_swapped` and `alibi_slopes` just like RoPE?

For now, I disabled `seqlenq_ngroups_swapped` when using ALiBi. I'm still looking for a way to use them together.
